README.md
# reactPointLock

[![npm download](https://img.shields.io/npm/dt/react-pointerlock.svg)](https://www.npmjs.com/package/react-pointerlock)
[![npm version](https://img.shields.io/npm/v/react-pointerlock.svg)](https://www.npmjs.com/package/react-pointerlock)
[![Build Status](https://img.shields.io/travis/jeongsd/react-pointerlock.svg)](https://travis-ci.org/jeongsd/react-pointerlock)
[![Licenese](https://img.shields.io/npm/l/react-pointerlock.svg)](./LICENSE)

The __Pointer Lock API__ (formerly called Mouse Lock API) provides input methods based on the movement of the mouse over time (i.e., deltas), not just the absolute position of the mouse cursor in the viewport. It gives you access to raw mouse movement, locks the target of mouse events to a single element, eliminates limits on how far mouse movement can go in a single direction, and removes the cursor from view. It is ideal for first person 3D games, for example.

example code using [MDN Pointer Lock API](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_Lock_API)

## installation
`npm install reactPointLock`


## Component API
__onMouseMove(movement, event)__
- movement.x : movement x point
- movement.y : movement y point
- event : mousemove event

__onPointLock()__
- when activate pointerlock

__onExitPointLock()__
- when user exit pointerlock ex) pressing ESC key, etc.

## Example
```js
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PointerLock from 'react-pointerlock';

import './app.css';

class App extends Component {

  constructor() {
    super();

    this.state = { x: 50, y: 50 };

    this.onMouseMove = this.onMouseMove.bind(this);
  }

  componentDidMount() {
    this.canvasDraw();
  }

  componentDidUpdate() {
    this.canvasDraw();
  }

  onMouseMove(movement) {
    const canvas = ReactDOM.findDOMNode(this.refs.canvas);
    let x = this.state.x + movement.x;
    let y = this.state.y + movement.y;

    if (x > canvas.clientWidth + 20) {
      x = 0;
    }

    if (y > canvas.clientHeight + 20) {
      y = 0;
    }

    if (x < -15) {
      x = canvas.clientWidth;
    }

    if (y < -15) {
      y = canvas.clientHeight;
    }

    this.setState({ x: x, y: y });
  }

  canvasDraw() {
    const canvas = ReactDOM.findDOMNode(this.refs.canvas);
    const context = canvas.getContext('2d');

    context.fillStyle = 'black';
    context.fillRect(0, 0, canvas.clientWidth, canvas.clientHeight);

    context.fillStyle = '#F9F903';
    context.beginPath();
    context.arc(this.state.x, this.state.y, 20, 0, Math.PI * 2, true);
    context.fill();
  }

  render() {
    return (
      <div className="app">
      <h1>Pointer lock demo</h1>
      <p>This demo demonstrates usage of the pointer lock API. Click on the canvas area and your mouse will directly control the ball inside the canvas, not your mouse pointer. You can press escape to return to the standard expected state.</p>

      <h2>X Position: { this.state.x }</h2>
      <h2>Y Position: { this.state.y }</h2>

      <PointerLock
        onMouseMove={ this.onMouseMove } >
        <canvas
          width="640px"
          height="369px"
          ref="canvas"/>
      </PointerLock>
      </div>
    );
  }

}

ReactDOM.render(
  <App />,
  document.getElementById('app')
);
```

## Contribute

```bash
npm install
npm start
```

```bash
npm build
```


## TODO
- test
- css build using babel.js
- minifly


CHANGELOG.md
#Change Logs


## 1.1.3 - Saturday, 12 December 07 : 35 : 39 (GMT 00:00)
- [7bd6566](../../commit/7bd6566) add onPointLock, onExitPointLock

## 1.1.2
- [bbffd9c](../../commit/bbffd9c) build format change commonjs -> [umd](https://github.com/rollup/rollup/wiki/JavaScript-API#format)  

## 1.1.1 - Monday, 07 December 13 : 10 : 19 (GMT 00:00)
- [faf122d](../../commit/faf122d) build script fix
- [faf122d](../../commit/faf122d) badge using [shields](https://github.com/badges/shields)
- [faf122d](../../commit/faf122d) add jsnext:main field [jsnext:main](https://github.com/rollup/rollup/wiki/jsnext:main)

## 1.1.0 - Monday, 07 December 11 : 45 : 56 (GMT 00:00)
- [6f3f214](../../commit/6f3f214) can't using build js file so build tools change [webpack](https://webpack.github.io/) -> [rollupjs](http://rollupjs.org/)


src/helper/pointLock.js

let pointerLockElement;
let pointerLockChange;
let requestPointerLock;
let exitPointerLock;
let pointerLockError;
let movementX;
let movementY;

if ('pointerLockElement' in document) {
  pointerLockElement = 'pointerLockElement';
  pointerLockChange = 'pointerlockchange';
  requestPointerLock = 'requestPointerLock';
  exitPointerLock = 'exitPointerLock';
  pointerLockError = 'pointerlockerror';
  movementX = 'movementX';
  movementY = 'movementY';
} else if ('mozPointerLockElement' in document) {
  pointerLockElement = 'mozPointerLockElement';
  pointerLockChange = 'mozpointerlockchange';
  requestPointerLock = 'mozRequestPointerLock';
  exitPointerLock = 'mozExitPointerLock';
  pointerLockError = 'mozpointerlockerror';
  movementX = 'mozMovementX';
  movementY = 'mozMovementY';
} else if ('webkitPointerLockElement' in document) {
  pointerLockElement = 'webkitPointerLockElement';
  pointerLockChange = 'webkitpointerlockchange';
  requestPointerLock = 'webkitRequestPointerLock';
  exitPointerLock = 'webkitExitPointerLock';
  pointerLockError = 'webkitpointerlockerror';
  movementX = 'webkitMovementX';
  movementY = 'webkitMovementY';
}

export const POINTERLOCK_ELEMENT = pointerLockElement;
export const POINTERLOCK_CHANGE = pointerLockChange;
export const REQUEST_POINTERLOCK = requestPointerLock;
export const EXIT_POINTERLOCK = exitPointerLock;
export const POINTERLOCK_ERROR = pointerLockError;
export const MOVEMENT_X = movementX;
export const MOVEMENT_Y = movementY;


src/react-pointerlock.js
import React from 'react';
import ReactDom from 'react-dom';
import classnames from 'classnames';
import { POINTERLOCK_ELEMENT, POINTERLOCK_CHANGE,
  REQUEST_POINTERLOCK, EXIT_POINTERLOCK, POINTERLOCK_ERROR,
  MOVEMENT_X, MOVEMENT_Y } from './helper/pointLock.js';
// import './react-pointerlock.css';

const propTypes = {
  className: React.PropTypes.string,
  onMouseMove: React.PropTypes.func,
  onPointLock: React.PropTypes.func,
  onExitPointLock: React.PropTypes.func,
  blockElement: React.PropTypes.node,
  // onError: PropTypes.func,
};

const defaultProps = {
  blockElement: (
    <div className="PointerLocker-blocker">
      <div className="PointerLocker-instructions">
        Click to View
      </div>
    </div>
  ),
};

class PointerLocker extends React.Component {

  constructor() {
    super();

    this.state = {
      isPointLock: false,
    };

    this.requestPointerLock = this.requestPointerLock.bind(this);
    this.onPointLockChange = this.onPointLockChange.bind(this);
    this.onMouseMove = this.onMouseMove.bind(this);
    this.onPointLockError = this.onPointLockError.bind(this);
    this.onClick = this.onClick.bind(this);
  }

  componentDidMount() {
    if (this.caniuse()) {
      document.addEventListener(POINTERLOCK_CHANGE, this.onPointLockChange, false);
      // document.addEventListener(POINTERLOCK_ERROR, this.onPointLockError, false);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.isPointLock !== prevState.isPointLock ) {
      if (this.state.isPointLock) {
        this.props.onPointLock();
      } else {
        this.props.onExitPointLock();
      }
    }
  }

  componentWillUnmount() {
    if (this.caniuse()) {
      document.removeEventListener(POINTERLOCK_CHANGE, this.onPointLockChange, false);
      // document.removeEventListener(POINTERLOCK_ERROR, this.onPointLockError, false);
    }
  }

  onPointLockChange() {
    const currentPointLockElement = document[POINTERLOCK_ELEMENT];
    const pointLockElement = ReactDom.findDOMNode(this.refs.pointerLocker);

    if (currentPointLockElement === pointLockElement) {
      // react onMouseMove event doesn't extension to mouse events
      // https://developer.mozilla.org/en-US/docs/Web/API/Pointer_Lock_API#extensions
      document.addEventListener('mousemove', this.onMouseMove, false);
      return this.setState({
        isPointLock: true,
      });
    }

    document.removeEventListener('mousemove', this.onMouseMove);
    this.setState({
      isPointLock: false,
    });
  }

  onClick() {
    this.requestPointerLock();
  }

  onMouseMove(event) {
    if (this.props.onMouseMove) {
      const movement = {
        x: event[MOVEMENT_X] || 0,
        y: event[MOVEMENT_Y] || 0,
      };

      this.props.onMouseMove(movement, event);
    }
  }

  onPointLockError(error) {
    console.error('requestPointerLock or exitPointerLock calling failed', error);

    if (this.props.onError) {
      this.props.onError(error);
    }
  }

  exitPointLock() {
    document[EXIT_POINTERLOCK]();
  }

  requestPointerLock() {
    const pointerLocker = ReactDom.findDOMNode(this.refs.pointerLocker);

    pointerLocker[REQUEST_POINTERLOCK]();
  }

  caniuse() {
    if (POINTERLOCK_ELEMENT) {
      return true;
    }

    return false;
  }

  render() {
    const className = classnames('PointerLocker', this.props.className);
    let blocker;

    if (!this.state.isPointLock) {
      blocker = this.props.blockElement;
    }

    return (
      <div
        ref="pointerLocker"
        className={ className }
        onMouseMove={
          this.state.isPointLock ? this.onMouseMove : null
        }
        onClick={ this.requestPointerLock } >
        { blocker }
        { this.props.children }
      </div>
    );
  }

}

PointerLocker.propTypes = propTypes;
PointerLocker.defaultProps = defaultProps;


export default PointerLocker;


example/app.js
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PointerLock from 'react-pointerlock';

import './app.css';

class App extends Component {

  constructor() {
    super();

    this.state = {
      x: 50,
      y: 50,
      isOnPointerLock: false,
    };

    this.onMouseMove = this.onMouseMove.bind(this);
    this.onPointLock = this.onPointLock.bind(this);
    this.onExitPointLock = this.onExitPointLock.bind(this);
  }

  componentDidMount() {
    this.canvasDraw();
  }

  componentDidUpdate() {
    this.canvasDraw();
  }

  onMouseMove(movement) {
    const canvas = ReactDOM.findDOMNode(this.refs.canvas);
    let x = this.state.x + movement.x;
    let y = this.state.y + movement.y;

    if (x > canvas.clientWidth + 20) {
      x = 0;
    }

    if (y > canvas.clientHeight + 20) {
      y = 0;
    }

    if (x < -15) {
      x = canvas.clientWidth;
    }

    if (y < -15) {
      y = canvas.clientHeight;
    }

    this.setState({ x: x, y: y });
  }

  onPointLock() {
    this.setState({
      isOnPointerLock: true,
    });
  }

  onExitPointLock() {
    this.setState({
      isOnPointerLock: false,
    });
  }

  canvasDraw() {
    const canvas = ReactDOM.findDOMNode(this.refs.canvas);
    const context = canvas.getContext('2d');

    context.fillStyle = 'black';
    context.fillRect(0, 0, canvas.clientWidth, canvas.clientHeight);

    context.fillStyle = '#F9F903';
    context.beginPath();
    context.arc(this.state.x, this.state.y, 20, 0, Math.PI * 2, true);
    context.fill();
  }

  render() {
    return (
      <div className="app">
      <h1>Pointer Lock Demo</h1>
      <p>This demo demonstrates usage of the pointer lock API. Click on the canvas area and your mouse will directly control the ball inside the canvas, not your mouse pointer. You can press escape to return to the standard expected state.</p>

      <h2>onPointLock? { this.state.isOnPointerLock ? 'on' : 'off'}</h2>
      <h2>X Position: { this.state.x }</h2>
      <h2>Y Position: { this.state.y }</h2>

      <PointerLock
        onPointLock={ this.onPointLock }
        onExitPointLock={ this.onExitPointLock }
        onMouseMove={ this.onMouseMove } >
        <canvas
          width="640px"
          height="369px"
          ref="canvas"/>
      </PointerLock>
      </div>
    );
  }

}

ReactDOM.render(
  <App />,
  document.getElementById('app')
);
